File: 1268-search-suggestions-system/1268-search-suggestions-system.java
class Solution {
    public List<List<String>> suggestedProducts(String[] products, String searchWord) {
        Trie root = new Trie();
        
        for(String p : products) {
            Trie head = root;
            for(char c : p.toCharArray()) {
                head = head.getNext(c);
                head.add(p);
            }
        }
        
        List<List<String>> res = new ArrayList<>();
        
        {
            Trie head = root;
            for(char c : searchWord.toCharArray()) {
                head = head.getNext(c);
                res.add(head.getWords());
            }
        }
        
        return res;
    }
    
    class Trie {
        public Set<String> words = new HashSet<>();
        public Map<Character, Trie> next = new HashMap<>();
        
        public Trie getNext(char c) {
            return next.computeIfAbsent(c, i -> new Trie());
        }
        
        public void add(String word) {
            words.add(word);
        }
        
        public List<String> getWords() {
            return words.stream().sorted().limit(3).collect(Collectors.toList());
        }
    }
}

File: 1268-search-suggestions-system/1268-search-suggestions-system.py
class Trie:
    def __init__(self):
        self.next = [None] * 26
        self.word = None
        
class Solution:
    root = None
    
    def suggestedProducts(self, products: List[str], searchWord: str) -> List[List[str]]:
        self.root = Trie()
        res = []
        
        self.buildTrie(products)
        
        current = self.root
        for c in searchWord:
            r = []
            
            if not current or not current.next[self.idx(c)]:
                current = None
                res.append(r)
                continue
            
            current = current.next[self.idx(c)]
            self.find(current, r)
            
            res.append(r)
            
        return res
        
        
    def find(self, n, res):
        if not n:
            return
        
        if len(res) >= 3:
            return
        
        if n.word:
            res.append(n.word)
            
        for i in range(26):
            if n.next[i]:
                self.find(n.next[i], res)
                if len(res) >= 3:
                    return
        
    
    def buildTrie(self, products):
        for prod in products:
            current = self.root
            for c in prod:
                if not current.next[self.idx(c)]:
                    current.next[self.idx(c)] = Trie()
                current = current.next[self.idx(c)]
            current.word = prod
            
                
    def idx(self, c):
        return ord(c) - ord('a')
        
